src/optimize_embedding_potential.F

```diff
@@ -928,6 +976,106 @@ SUBROUTINE release_opt_embed(opt_embed)
+      REAL(KIND=dp)                                      :: dvol, normalize_factor
+      REAL(KIND=dp), DIMENSION(:), POINTER               :: rhs_subsys
+      TYPE(particle_list_type), POINTER                  :: particles
+      TYPE(pw_env_type), POINTER                         :: pw_env
+      TYPE(pw_p_type)                                    :: rho_resp, v_resp_gspace, v_resp_rspace
+      TYPE(pw_poisson_type), POINTER                     :: poisson_env
+      TYPE(pw_pool_type), POINTER                        :: auxbas_pw_pool
+      TYPE(qs_subsys_type), POINTER                      :: subsys
+
+      ! Get available particles
+      NULLIFY (subsys)
+      CALL get_qs_env(qs_env=qs_env, subsys=subsys, pw_env=pw_env)
+      CALL qs_subsys_get(subsys, particles=particles)
+      natom = particles%n_els
+
+      ALLOCATE (rhs_subsys(natom))
+
+      NULLIFY (map_index)
+      CALL get_subsys_map_index(mapping_section, natom, iforce_eval, nforce_eval, &
+                                map_index, .TRUE.)
+
+      ! Mapping particles from iforce_eval environment to the embed env
+      DO iparticle = 1, natom
+         jparticle = map_index(iparticle)
+         rhs_subsys(iparticle) = rhs(jparticle)
+      END DO
+
+      ! Prepare plane waves
+      NULLIFY (auxbas_pw_pool)
+
+      CALL pw_env_get(pw_env, auxbas_pw_pool=auxbas_pw_pool, &
+                      poisson_env=poisson_env)
+
+      CALL pw_pool_create_pw(auxbas_pw_pool, &
+                             v_resp_gspace%pw, &
+                             use_data=COMPLEXDATA1D, &
+                             in_space=RECIPROCALSPACE)
+
+      CALL pw_pool_create_pw(auxbas_pw_pool, &
+                             v_resp_rspace%pw, &
+                             use_data=REALDATA3D, &
+                             in_space=REALSPACE)
+
+      CALL pw_pool_create_pw(auxbas_pw_pool, rho_resp%pw, &
+                             use_data=REALDATA3D, &
+                             in_space=REALSPACE)
+
+      ! Calculate charge density
+      CALL pw_zero(rho_resp%pw)
+      CALL calculate_rho_resp_all(rho_resp, rhs_subsys, natom, eta, qs_env)
+
+      ! Calculate potential
+      CALL pw_zero(v_resp_gspace%pw)
+      CALL pw_poisson_solve(poisson_env, rho_resp%pw, &
+                            vhartree=v_resp_gspace%pw)
+      CALL pw_zero(v_resp_rspace%pw)
+      CALL pw_transfer(v_resp_gspace%pw, v_resp_rspace%pw)
+      dvol = v_resp_rspace%pw%pw_grid%dvol
+      CALL pw_scale(v_resp_rspace%pw, dvol)
+      normalize_factor = SQRT((eta/pi)**3)
+      !normalize_factor = -2.0_dp
+      CALL pw_scale(v_resp_rspace%pw, normalize_factor)
+
+      ! Hard copy potential
+      v_rspace%pw%cr3d(:, :, :) = v_resp_rspace%pw%cr3d(:, :, :)
+
+      ! Release plane waves
+      CALL pw_release(v_resp_gspace%pw)
+      CALL pw_release(v_resp_rspace%pw)
+      CALL pw_release(rho_resp%pw)
+
+      ! Deallocate map_index array
+      IF (ASSOCIATED(map_index)) THEN
```

Comment:
the `IF()` here is rather rhetorical, there is no path in which `map_index` is not associated, or am I missing something?

Reply:
Yes, and thus superfluent.